tests/unit/parsers/test.manifestjson.js

```diff
@@ -192,24 +192,36 @@ describe('ManifestJSONParser', () => {
   });
 
   describe('bad permissions', () => {
-    it('should not error if permission is a string (even if unknown)', () => {
-      const addonLinter = new Linter({ _: ['bar'] });
-      const json = validManifestJSON({
-        permissions: ['identity', 'fileSystem'],
-      });
-
-      const manifestJSONParser = new ManifestJSONParser(
-        json,
-        addonLinter.collector
-      );
-      expect(manifestJSONParser.isValid).toEqual(false);
-      const { warnings } = addonLinter.collector;
-      expect(addonLinter.collector.errors.length).toBe(0);
-      assertHasMatchingError(warnings, {
-        code: messages.MANIFEST_PERMISSIONS.code,
-        message: /Unknown permissions "fileSystem"/,
+    for (const manifestKey of ['permissions', 'optional_permissions']) {
+      const expectedMsgCode =
+        manifestKey === 'permissions'
+          ? messages.MANIFEST_PERMISSIONS.code
+          : messages.MANIFEST_OPTIONAL_PERMISSIONS.code;
+      const json =
+        manifestKey === 'permissions'
+          ? validManifestJSON({
+              permissions: ['identity', 'fileSystem'],
+            })
+          : validManifestJSON({
+              optional_permissions: ['idle', 'fileSystem'],
+              applications: { gecko: { strict_max_version: '55.0' } },
+            });
+      it(`should not error if ${manifestKey} is a string (even if unknown)`, () => {
+        const addonLinter = new Linter({ _: ['bar'] });
+        const manifestJSONParser = new ManifestJSONParser(
+          json,
+          addonLinter.collector
+        );
+        expect(manifestJSONParser.isValid).toEqual(false);
+        const { warnings } = addonLinter.collector;
+        console.log({ warnings });
+        expect(addonLinter.collector.errors.length).toBe(0);
+        assertHasMatchingError(warnings, {
+          code: expectedMsgCode,
+          message: `/Unknown ${manifestKey} "fileSystem"/`,
+        });
       });
-    });
+    }
 
     it('should error if permission is not a string', () => {
```

Comment:
This `should error if permission is not a string` looks like another test that we may want to also run for the optional_permissions manifest key (and reuse the same for look if there are multiple tests in the same describe group that we have to run on both permissions and optional_permissions).

Also the test that follows, `should error if permission is duplicated`, seems also something that we may want to run on the optional_permissions (and it may require some other changes to pass for the optional_permissions, because none of the implementation currently changes in this PR seems to be covering linting errors related to duplicated manifest keys and so I would expect the test to fail on the optional_permissions until we look into that and make sure that the same thing is enforced on optional_permissions as it is at the moment for the permissions manifest key).

Reply:
`optional_permissions` doesn't have MANIFEST_BAD_PERMISSION error, should optional_permissions have a such type of error tho?
Or should it have a conditional statement for a warning?
Also currently there's no warning for duplicates in `optional_permissions` and for a "not a string value" it gives the following warning 
```
Code                            Message                   Description                                      File            Line   Column
MANIFEST_OPTIONAL_PERMISSIONS   /optional_permissions:    See https://mzl.la/2Qn0fWC (MDN Docs) for more   manifest.json                
                                Unknown                   information.                                                                  
                                optional_permissions                                                                                    
                                "[object Object]" at 2.  
```

Reply:
> `optional_permissions` doesn't have MANIFEST_BAD_PERMISSION error, should optional_permissions have a such type of error tho?
> Or should it have a conditional statement for a warning?

We don't currently collect a `MANIFEST_BAD_PERMISSION` error for the optional_permissions because at the moment the related code in `src/parser/manifestjson.js` is only executed when `error.dataPath.startsWith('/permissions')` and so we are currently skipping that check for the `optional_permissions`:
https://github.com/mozilla/addons-linter/blob/5b4399ed3117bc7a8759b5ff5d24f1aa0a4cd1ca/src/parsers/manifestjson.js#L228-L234

Nevertheless, there shouldn't be any reason for that check to not be extended to the `optional_permissions` (eventually by also defining a new `MANIFEST_BAD_OPTIONAL_PERMISSION` so that we can have a different shorten url in the message description).

> Also currently there's no warning for duplicates in `optional_permissions` and for a "not a string value" it gives the following warning

I took a look into where the duplicated entries are detected for the `permissions` manifest property and I'm pretty sure that it is covered by the `uniqueItems` property here:

https://github.com/mozilla/addons-linter/blob/7530c81c27e7d8484b7d76381924489eaf07a5b3/src/schema/imported/manifest.json#L221-L241

You can double-check that it is actually the case by removing (or flipping the value) that `uniqueItems` property and run the test case that you think it should then fail.

To cover this for the optional permissions we should make sure that the entries for the `optional_permissions` in the same file is also going to contain that `uniqueItems` property.

This unfortunately is not enough, because we do update all those schema files in the `src/schema/imported` on every new beta release, and if we do only add the `uniqueItems` property to `src/schema/imported/manifest.json` then we will lose it once we import the schema for Firefox 81 beta.

To make sure that the `uniqueItems` property is going to be preserved over all the next schema imports, we do also need to apply the same kind of change we applied to `src/schema/updates/manifest.json` for the `permissions` property here:
https://github.com/mozilla/addons-linter/blob/6b26308027569cb674350ae30417cf74a878ff99/src/schema/updates/manifest.json#L40-L48

the schemas in `src/schema/updates/` are basically describing the changes to apply on the schema files as imported from Firefox, and so by applying the change to `src/schema/updates/manifest.json` we also ensure that the `uniqueItems` will be preserved when we will import new versions of the schema files from Firefox.

Let me know if some parts of this comment are not clear enough, otherwise give it a try and let me know how it goes (and feel free to update this PR even if there are still issues to figure out and solve, I'll be happy to take a look to the drafted changes, and travis logs if any test does still fail, and provide more guidance based on that)